docs/configuration.md
# Configuration

Each saffron instance need a configuration file. Below is the structure of the configuration file.

## `mode`
Default value: `main`

Can receive the values `main` and `worker`. It specifies if the saffron instance will be the main instance
or a worker only instance.

If the environment variable `SAFFRON_MODE` is set properly it will override this field.

## `newArticles`
Default value: `none`

A function that will be called after saffron is done scraping and parsing the articles.
This function is useful to save the parsed articles in the database.

Arguments:
* `tableName`: The table (or collection) where the articles will be saved.
* `articles`: An array of all the retrieved articles.

## `sources`

### `path`
Default value: `./sources`

The directory where the source file are located.

All `.js` and `.json` files will be treated as source files (including subdirectories)

### `scanSubFolders`
Default value: `true`

If `true` the Saffron will scan all the sub directories inside the `path` directory.

### `dynamicSourceFiles`
An array containing all the implementations for the dynamic source files.

### `loader`
Default value: `JSON.parse(fs.readFileSync(filepath))`

A custom loader that will allow to manually load each file to Saffron.

It was created to allow ES6 projects to load javascript files for the old dynamic parser.
Now it can be used to preprocess the files before passing them to Saffron.

```typescript
loader: async (filepath: string) => {
    let data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
    // Proccess the content of the source file
    // ...
    return data;
}
```

### `includeOnly`
Default value: `[]`

When not empty it will filter the parsed sources and include only the ones that are mentioned.
The source identification is made using the source name (and not the source file name).

This field is useful when testing specific source files.

### `exclude`
Default value: `[]`

When not empty it will exclude the parsed sources and include only the ones that are not mentioned here.
It will exclude source files even if they are in [`includeOnly`](#includeonly) option.

The source identification is made using the source name (and not the source file name).

This field is useful when a source file is deprecated and must be excluded.
Setting a environment variable for this may be a good choice.

## `workers`

### `nodes`
Default value: `1`

The workers that will be initialized. All workers run in the same instance.
There is not a clear benefit in having multiple workers, they will not
run in different threads or processes. They exist for purpose of analytics.

If you want to run saffron in different processes check out about [`Grid`](#grid).

If you want to assign a name to each worker, you can replace the number with
an array of names (strings), like: `nodes: ['Worker 1', 'Worker 2', 'Billy']`.

### `delayBetweenRequests`
Default value: `0`

There is a case where a website may rate limit the requests it accepts, so a source
file with multiple urls may fail due to this limit.

Increasing the option `delayBetweenRequests` (in milliseconds) will put a distance
between the request that belong in the same source file.

### `axios`
Axios' configuration that will be applied to the requests made by saffron.
The method of the requests is by default `GET`, but it can be overridden here.

It supports synchronous callback like:

```typescript
axios: async (source: Source) => {
    return {
        method: 'POST',
        data: {
            key: 'value'
        },
        timeout: 3000
    };
}
```

### `preprocessor`
A function that will edit the response of the http request before
passing it to the parser.

It is helpful in many cases where modifications are needed to the document
before parsing, for example an RSS file which contains the character `&`
instead of the encoded representation `&amp;`.

It supports synchronous callback like:

```typescript
// type RequestsResult = AxiosResponse | AxiosResponse[];
axios: async (responses: RequestsResult, source: Source) => {
    // ...
    return responses;
}
```

This function does apply to the `dynamic` parser.

### `articles.amount`
Default value: `30`

The maximum amount of articles that saffron will return for each source scraping job.

### `articles.includeContentAttachments`
Default value: `true`

If the generated article's attachments will contain the extracted urls from article's content.

### `articles.includeCategoryUrlsIn`
Default value: `categories`

Accepted values are `categories` or `extras`.
THis field will specify where to store the categories mentioned in the [`url`](./source_files/source_file.md#url) field of the source file.

## `scheduler`

### `jobsInterval`
Default value: `3600000` (60 minutes)

The interval in milliseconds, between each source scraping job for a specific source.

This option is also used by the scheduler to spread the source files evenly inside
a specific timeframe. This will help to ease your machine by not executing all the
source scraping jobs together.

A lower value for this interval means a heavier load on the websites that are scraped.
This field has a recommended minimum value of `5000ms`.

### `heavyJobFailureInterval`
Default value: `86400000` (24 hours)

When a source scraping job fails consecutively 10 times in a row it wil trigger
the freeze mechanism of saffron and schedule the source to be executed again after
`heavyJobFailureInterval` milliseconds.

It is recommended to have a high value for this field as it will be used when a website is down.

### `noResponseThreshold`
Default value: `2`

A multiplier of the source interval. When exceeding that product it will re-assign a worker.

For example for an `interval` of 1 minute and `noResponseThreshold` of 2, if a worker does
not report the source scraping job as passed or failed after `interval` * `noResponseThreshold` =
2 minutes, it will replace the current worker.

This field is useful in cases where a remote node has crashed or lost connection with main instance.

### `randomizeInterval`
Default value: `-300 to 300 seconds`

The scheduler adds/subtracts a random time from the interval of each job to avoid making
requests at fixed intervals. The result of the function must be in milliseconds.

```javascript
randomizeInterval: () => {
    const random = Math.floor(Math.random() * (high - low) + low) * 1000;
    return Math.random() >= 0.5 ? random : -random;
}
```

## `grid`
### `distributed`
Default value: `false`

In the `main` instance it will start a `socket.io-server` and wait for nodes to be connected.
In case of a `worker` instance it will try to connect to the `main` by using `socket.io-client`.

### `useHTTPS`
Default value: `false`

Use HTTPS instead of HTTP server.

### `serverHost`
The server's host where the `worker` nodes will connect to.

Without the `http://` or `https://` prefix and `:<port>` postfix.

### `serverPort`
Default value: `3000`

The port where the `main` node will listen and `worker` nodes will connect

### `authToken`
A string that will authenticate a `worker` node to the `main` node.

if the authentication fails it will disconnect the socket.

### `key`
The key used by the HTTPS server.

### `cert`
The certificate used by the HTTPS server.

## Misc
### `log`
Default value: `all`

The log level of saffron.

Each level is a subset of the next one: `none` &lt; `errors` &lt; `info` &lt; `all`  

### `eventDelay`
Default value: `0`

The time in milliseconds where the events will be delayed before delivered.

# Environment
Saffron supports different configuration based on the environment.

The configuration will first receive the root configuration:
```typescript
const config = {
    newArticles: (tableName: string, articles: Article[]) => {
        // push to production database
    },
    // ...
    misc: {
        log: "errors" // log only errors
    },
    // ...
};
```

and then read the child object `development`, `production` or `testing` based on the environment:

```typescript
const config = {
    newArticles: (tableName: string, articles: Article[]) => {
        // push to production database
    },
    // ...
    misc: {
        log: "errors" // log only errors
    },
    // ...
    development: {
        newArticles: (tableName: string, articles: Article[]) => {
            // push to local database
        },
        // ...
        misc: {
            log: "all" // log everything
        },
        // ...
    }
};
```


src/components/config.ts
import fs from "fs";
import type {ConfigType, MergedConfig} from "./types";
import _ from "lodash"

export enum ConfigOptions {
    SOURCES_PATH = 0,
    SOURCES_INCLUDE_ONLY = 1,
    SOURCES_EXCLUDE = 2,
    MODE = 3,
    WORKER_NODES = 4,
    ARTICLE_AMOUNT = 5,
    JOB_INT = 6,
    JOB_HEAVY_INT = 7,
    NO_RESPONSE_THR = 8,
    INT_RANDOMIZER = 9,
    PREPROCESSOR = 10,
    DISTRIBUTED = 13,
    HOST = 14,
    PORT = 15,
    AUTH_TOKEN = 16,
    USE_HTTPS = 17,
    HTTPS_KEY = 18,
    HTTPS_CERT = 19,
    LOG_LEVEL = 20,
    EVENT_DELAY = 21,
    NEW_ARTICLES = 22,
    INCLUDE_CNT_ATTACHMENTS = 23,
    INCLUDE_CAT_URL = 24,
    AXIOS_REQUEST_CONFIG = 25,
    SCAN_SUB_FOLDERS = 26,
    SOURCE_LOADER = 27,
    DELAY_BETWEEN_REQUESTS = 29,
    DYNAMIC_SOURCE_FILES = 30,
}

const defaultConfig: ConfigType = {
    mode: "main",
    newArticles: (tableName, articles) => {
    },
    sources: {
        path: "./sources",
        scanSubFolders: true,
        dynamicSourceFiles: [],
        loader: (filepath: string) => JSON.parse(fs.readFileSync(filepath, 'utf-8')),
        includeOnly: [],
        exclude: []
    },
    workers: {
        nodes: 1, // Start one worker
        delayBetweenRequests: 0,
        axios: {
            timeout: 10000,
            headers: {},
            maxRedirects: 5
        },
        preprocessor: async (r, s) => r,
        articles: {
            amount: 30,
            includeContentAttachments: true,
            includeCategoryUrlsIn: 'categories'
        }
    },
    scheduler: {
        jobsInterval: 3600000,
        heavyJobFailureInterval: 86400000,
        noResponseThreshold: 2,
        randomizeInterval: () => {
            if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'testing')
                return 0;

            const high = 300;
            const low = 0;

            const random = Math.floor(Math.random() * (high - low) + low) * 1000;
            return Math.random() >= 0.5 ? random : -random;
        }
    },
    grid: {
        distributed: false,
        useHTTPS: false,
        serverHost: '127.0.0.1',
        serverPort: 3000
    },
    misc: {
        log: "all",
        eventDelay: 0
    }
};

export class Config {
    config: ConfigType = _.cloneDeep(defaultConfig);

    constructor(config?: MergedConfig) {
        this.initializeConfig(config);
    }

    static getOption(option: ConfigOptions, config: Config | null): any {
        const conf = config?.config ?? defaultConfig;

        switch (option) {
            case ConfigOptions.MODE:
                return conf.mode;

            case ConfigOptions.SOURCES_PATH:
                return conf.sources?.path;
            case ConfigOptions.SCAN_SUB_FOLDERS:
                return conf.sources?.scanSubFolders;
            case ConfigOptions.SOURCES_INCLUDE_ONLY:
                return conf.sources?.includeOnly;
            case ConfigOptions.SOURCES_EXCLUDE:
                return conf.sources?.exclude;
            case ConfigOptions.SOURCE_LOADER:
                return conf.sources?.loader;
            case ConfigOptions.DYNAMIC_SOURCE_FILES:
                return conf.sources?.dynamicSourceFiles;

            case ConfigOptions.WORKER_NODES:
                return conf.workers?.nodes;
            case ConfigOptions.DELAY_BETWEEN_REQUESTS:
                return conf.workers?.delayBetweenRequests;
            case ConfigOptions.AXIOS_REQUEST_CONFIG:
                return conf.workers?.axios
            case ConfigOptions.PREPROCESSOR:
                return conf.workers?.preprocessor

            case ConfigOptions.ARTICLE_AMOUNT:
                return conf.workers?.articles?.amount;
            case ConfigOptions.INCLUDE_CNT_ATTACHMENTS:
                return conf.workers?.articles?.includeContentAttachments;
            case ConfigOptions.INCLUDE_CAT_URL:
                return conf.workers?.articles?.includeCategoryUrlsIn;

            case ConfigOptions.JOB_INT:
                return conf.scheduler?.jobsInterval;
            case ConfigOptions.JOB_HEAVY_INT:
                return conf.scheduler?.heavyJobFailureInterval;
            case ConfigOptions.INT_RANDOMIZER:
                return conf.scheduler?.randomizeInterval;
            case ConfigOptions.NO_RESPONSE_THR:
                return conf.scheduler?.noResponseThreshold;

            case ConfigOptions.DISTRIBUTED:
                return conf.grid?.distributed;
            case ConfigOptions.HOST:
                return conf.grid?.serverHost;
            case ConfigOptions.PORT:
                return conf.grid?.serverPort;
            case ConfigOptions.AUTH_TOKEN:
                return conf.grid?.authToken;
            case ConfigOptions.USE_HTTPS:
                return conf.grid?.useHTTPS;
            case ConfigOptions.HTTPS_KEY:
                return conf.grid?.key;
            case ConfigOptions.HTTPS_CERT:
                return conf.grid?.cert;

            case ConfigOptions.LOG_LEVEL:
                return conf.misc?.log;
            case ConfigOptions.EVENT_DELAY:
                return conf.misc?.eventDelay;
            case ConfigOptions.NEW_ARTICLES:
                return conf.newArticles;
        }
    }

    initializeConfig(config?: MergedConfig) {
        // To avoid overriding user's object
        config = _.cloneDeep(config);

        this.config = this.mergeObject(config, this.config);

        switch (process.env.NODE_ENV) {
            case "production":
                if ((config as any).production)
                    this.config = this.mergeObject(config!.production, this.config);
                break;
            case "development":
                if ((config as any).development)
                    this.config = this.mergeObject(config!.development, this.config);
                break;
            case "testing":
                if ((config as any).testing)
                    this.config = this.mergeObject(config!.testing, this.config);
                break;
        }

        if (process.env.SAFFRON_MODE && ["main", "worker"].includes(process.env.SAFFRON_MODE))
            this.config.mode = <any>process.env.SAFFRON_MODE;

        if (this.config.mode !== 'main' && this.config.mode !== 'worker')
            throw new Error('ConfigurationException Option mode is not valid, requirements(type = string, =main or =worker)');

        if (typeof this.config.newArticles !== 'function')
            throw new Error('ConfigurationException Option newArticles is not valid, requirements(type = function)');

        if (typeof this.config.sources !== 'object' || Array.isArray(this.config.sources))
            throw new Error('ConfigurationException Option block sources is not valid, requirements(type = object)');
        if (typeof this.config.sources.path !== 'string')
            throw new Error('ConfigurationException Option sources.path is not valid, requirements(type = string)');
        if (typeof this.config.sources.scanSubFolders !== 'boolean')
            throw new Error('ConfigurationException Option sources.scanSubFolders is not valid, requirements(type = boolean)');
        if (!Array.isArray(this.config.sources.includeOnly))
            throw new Error('ConfigurationException Option sources.includeOnly is not valid, requirements(type = string-array)');
        if (!Array.isArray(this.config.sources.exclude))
            throw new Error('ConfigurationException Option sources.exclude is not valid, requirements(type = string-array)');

        if (typeof this.config.workers !== 'object' || Array.isArray(this.config.workers))
            throw new Error('ConfigurationException Option block workers is not valid, requirements(type = object)');
        if (typeof this.config.workers.nodes === 'number') {
            if (this.config.workers.nodes < 0)
                throw new Error('ConfigurationException Option workers.nodes is not valid, requirements(type = number, positive or zero)');
        } else if (Array.isArray(this.config.workers.nodes)) {
            const nodes = this.config.workers.nodes
            if ((new Set(nodes)).size !== nodes.length)
                throw new Error("worker.nodes cannot have duplicates names");
        } else
            throw new Error('ConfigurationException Option workers.nodes is not valid, requirements(type = string-array or number)');
        if (typeof this.config.workers.delayBetweenRequests !== 'number' || this.config.workers.delayBetweenRequests < 0)
            throw new Error('ConfigurationException Option workers.delayBetweenRequests is not valid, requirements(type = number, positive or zero)');
        if (typeof this.config.workers.articles !== 'object' || Array.isArray(this.config.workers.articles))
            throw new Error('ConfigurationException Option block workers.articles is not valid, requirements(type = object)');
        if (typeof this.config.workers.articles.amount !== 'number' || this.config.workers.articles.amount <= 0)
            throw new Error('ConfigurationException Option workers.articles.amount is not valid, requirements(type = number, positive)');
        if (typeof this.config.workers.articles.includeContentAttachments !== 'boolean')
            throw new Error('ConfigurationException Option workers.articles.includeContentAttachments is not valid, requirements(type = boolean)');
        if (typeof this.config.workers.articles.includeCategoryUrlsIn !== 'undefined'
            && this.config.workers.articles.includeCategoryUrlsIn !== 'categories'
            && this.config.workers.articles.includeCategoryUrlsIn !== 'extras'
        )
            throw new Error('ConfigurationException Option workers.articles.includeCategoryUrlsIn is not valid, requirements(type = string, =categories, =extras)');

        if (typeof this.config.scheduler !== 'object' || Array.isArray(this.config.scheduler))
            throw new Error('ConfigurationException Option block scheduler is not valid, requirements(type = object)');
        if (typeof this.config.scheduler.jobsInterval !== 'number')
            throw new Error('ConfigurationException Option scheduler.jobsInterval is not valid, requirements(type = number, >= 5000)');
        if (typeof this.config.scheduler.heavyJobFailureInterval !== 'number' || this.config.scheduler.heavyJobFailureInterval <= 0)
            throw new Error('ConfigurationException Option scheduler.heavyJobFailureInterval is not valid, requirements(type = number, positive)');
        if (typeof this.config.scheduler.noResponseThreshold !== 'number' || this.config.scheduler.noResponseThreshold <= 0)
            throw new Error('ConfigurationException Option scheduler.noResponseThreshold is not valid, requirements(type = number, positive)');
        if (typeof this.config.scheduler.randomizeInterval !== 'function')
            throw new Error('ConfigurationException Option scheduler.randomizeInterval is not valid, requirements(type = function)');

        if (typeof this.config.grid !== 'object' || Array.isArray(this.config.grid))
            throw new Error('ConfigurationException Option block grid is not valid, requirements(type = object)');
        if (typeof this.config.grid.distributed !== 'boolean')
            throw new Error('ConfigurationException Option grid.distributed is not valid, requirements(type = boolean)');
        if (this.config.grid.distributed) {
            if (typeof this.config.grid.useHTTPS !== 'boolean')
                throw new Error('ConfigurationException Option grid.useHTTPS is not valid, requirements(type = boolean)');
            if (typeof this.config.grid.serverHost !== 'string')
                throw new Error('ConfigurationException Option grid.serverHost is not valid, requirements(type = string)');
            if (typeof this.config.grid.serverPort !== 'number' || this.config.grid.serverPort < 1 || this.config.grid.serverPort > 65535)
                throw new Error('ConfigurationException Option grid.serverPort is not valid, requirements(type = number, 1..65535)');
            if (typeof this.config.grid.authToken !== 'string')
                throw new Error('ConfigurationException Option grid.authToken is not valid, requirements(type = string)');
        }

        if (typeof this.config.misc !== 'object' || Array.isArray(this.config.misc))
            throw new Error('ConfigurationException Option block misc is not valid, requirements(type = object)');
        if (!'all,info,errors,none'.split(',').includes(this.config.misc.log!))
            throw new Error('ConfigurationException Option misc.log is not valid, requirements(type = string, =all or =info or =errors or =none)');
        if (typeof this.config.misc.eventDelay !== 'number' || this.config.misc.eventDelay < 0)
            throw new Error('ConfigurationException Option misc.eventDelay is not valid, requirements(type = number, positive or zero)');
    }

    private mergeObject(src: any, original: object): any {
        return _.mergeWith({}, original, src, (o, s) => {
            if (typeof o == 'object' && !Array.isArray(o) && typeof s != 'function' && typeof s != 'function')
                return this.mergeObject(s, o);
            return s != null ? s : o;
        });
    }
}